File: backend/app/services/metrics_store.py
from datetime import datetime

from app.models.metrics import MetricsUpdate, SystemMetrics


class MetricsStore:
    """In-memory store for system metrics."""

    def __init__(self):
        self._systems: dict[str, SystemMetrics] = {}
        self._last_update: datetime | None = None

    def update(self, system_id: str, metrics: SystemMetrics) -> None:
        """Update metrics for a specific system."""
        self._systems[system_id] = metrics
        self._last_update = datetime.utcnow()

    def get(self, system_id: str) -> SystemMetrics | None:
        """Get metrics for a specific system."""
        return self._systems.get(system_id)

    def get_all(self) -> dict[str, SystemMetrics]:
        """Get all system metrics."""
        return self._systems.copy()

    def remove(self, system_id: str) -> None:
        """Remove a system from the store."""
        self._systems.pop(system_id, None)

    def clear(self) -> None:
        """Clear all metrics."""
        self._systems.clear()
        self._last_update = None

    def create_update_message(self) -> MetricsUpdate:
        """Create a metrics update message for WebSocket broadcast."""
        return MetricsUpdate(
            type="metrics_update",
            timestamp=datetime.utcnow(),
            systems=self._systems.copy(),
        )


metrics_store = MetricsStore()

File: backend/src/services/metrics_store.rs
use std::collections::HashMap;
use std::sync::Arc;

use tokio::sync::RwLock;

use crate::models::{MetricsUpdate, SystemMetrics};

#[derive(Clone, Default)]
pub struct MetricsStore {
    systems: Arc<RwLock<HashMap<String, SystemMetrics>>>,
}

impl MetricsStore {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn update(&self, system_id: &str, metrics: SystemMetrics) {
        self.systems
            .write()
            .await
            .insert(system_id.to_string(), metrics);
    }

    pub async fn get(&self, system_id: &str) -> Option<SystemMetrics> {
        self.systems.read().await.get(system_id).cloned()
    }

    pub async fn get_all(&self) -> HashMap<String, SystemMetrics> {
        self.systems.read().await.clone()
    }

    pub async fn clear(&self) {
        self.systems.write().await.clear();
    }

    pub async fn create_update_message(&self) -> MetricsUpdate {
        MetricsUpdate::new(self.systems.read().await.clone())
    }
}
